### ISSUE TITLE
Implement Bitswap protocol using notifications protocol

### ISSUE THREAD

**Issue by altonen** [paritytech/polkadot-sdk#524]
Bitswap protocol cannot be served over libp2p's request response protocol because it is not compatible with how go-ipfs and js-ipfs work. They seem to expect the response to be received in a new substream.

After paritytech/substrate#5685 has been implemented, rewrite the Bitswap protocol using notifications protocol and use the Bitswap query result as the handshake of the stream as is done in [`ipfs-embed`'s implementation](https://github.com/ipfs-rust/libp2p-bitswap/blob/master/src/compat/protocol.rs#L61-L65)

Related to paritytech/substrate#7915

**Comment by burdges** [comment]
I donno how we'd use bitswap, but if we want to see how fast availability can go, then we could try libtorrent on a testnet sometime.  It'd be a lot of work to do this securely,  but maybe interesting regardless. 

Anyways I asked this:  https://security.stackexchange.com/questions/271968/hardened-bittorrent-dos-vulnerabilities

**Comment by altonen** [comment]
I agree that it would be interesting to try libtorrent and see what kind of improvements we'd get out of it. Hopefully some time next year.

**Comment by tomaka** [comment]
> Anyways I asked this: https://security.stackexchange.com/questions/271968/hardened-bittorrent-dos-vulnerabilities

I don't understand the purpose of your question.
To me, the entire point of using bitswap or bittorent is compatibility with existing IPFS/Bittorent clients. If compatibility with existing software isn't required, then we're much better off using our own protocol.

**Comment by burdges** [comment]
> we're much better off using our own protocol rather than modify something existing.

Arvid Nordberg's libtorrent is famous for being extremely fast.  libp2p has a fairly poor reputation, no?

I didn't say we should switch, or even that we do this securely, or that doing this should be high priority, but doing this insecurely for a benchmark is a reasonable experiment that sets our performance goals.

My question was about how hard it is to do this securely, not immediately relevant. but interesting.

**Comment by tomaka** [comment]
>  libp2p has a fairly poor reputation, no?

That's a complicated question. We've had performance issues in the Substrate networking but that were due to Substrate-specific issues. We've had performance issues due to TCP slow start, which indeed could be solved by uTP, but it's not fair to attribute this to libp2p itself.

In terms of software, I feel like introducing a completely new library in the tech stack adds a lot of complexity (which Substrate is already packed with), and while it works as a short term hack in the long term it would be a better idea to instead improve libp2p.

**Comment by altonen** [comment]
We've had quite a bit of problems lately with libp2p. Both in terms of CPU usage which hinder the ability to scale the number of validators and parachains and also with these libp2p upgrades. Firstly they take a lot of effort to make, review and test and I feel like it's for little benefit because libp2p more or less works but if we don't keep up with their release schedule, the next upgrade we have to do is very painful.

In fact, latest release broke both Kademlia and syncing for us and we had to revert and probably have to get back upgrading sooner or later again. This is a lot of work that I would rather allocate somewhere else, such as even testing how well libtorrent would work, even if we find it unsuitable for whatever reason.

I agree with the point that adding a new library will add complexity but if in testing we find it so good that it would be stupid not to use it, I think we're able to manage the complexity of the additional library with abstractions. Not easy, but doable.

**Comment by tomaka** [comment]
> This is a lot of work that I would rather allocate somewhere else, such as even testing how well libtorrent would work

What I'm saying is that testing libtorrent is not an alternative to the work on libp2p, it comes *in addition* to the work you need to do on libp2p anyway. Unless you're thinking about replacing the entire networking stack with libtorrent, which is not realistic.

**Comment by altonen** [comment]
I'm not suggesting replacing the entire networking stack with libtorrent, that's probably not even possible in practice.

What I am suggesting is that we could introduce libtorrent as an additional transport to the other transport that we have right now and exposing that transport as a subset of protocols to the upper-layer protocols of Polkadot. This additional transport and whatever protocols it supports could still be advertised over `libp2p-identify`, allowing, for example, two validators to discover that they use libtorrent and switch to use that transport instead of using notifications/request-response. This is very high-level idea and I'm sure there are a ton of implementation details that prove out to be difficult implement but I don't see any fundamental problem with the approach.

**Comment by burdges** [comment]
@altonen  We never need protocol negotiation between validators afaik.  We make them upgrade their host or kick them from the validator set.

Imho there is considerable work in using libtorrent securely in availability, at which I hint in my SE question.  You need (1) some handle over its network layer encryption, (2) a representation of the torrent compatible with the candidate receipt, including (3) the chunks being files with appropriate prioritization for either your own chunk in availability or the systematic chunks in approvals, and (4) a local replacement for the tracker which builds the tracker data from availability bitfields plus backing & approvals statements.  Also (4) breaks real bittorrent compatibility, so you've not added functionality, just replaces the availability transport.

We'd only need some hacky (3) if we only ran this for performance tests though.  We'd learn something important:  If libtorrent offers no improvement, then we should stop blaming networking and look elsewhere for performance gains.  If libtorrent provides some improvement, like maybe less CPU even if LEDBAT adds latency, then we've still learned something about our networking code, and possible solutions.

**Comment by tomaka** [comment]
As I've mentioned above, I agree that rust-libp2p has many drawbacks (some of which being that nobody at Parity seems to be familiar with its internals anymore (and I'm not either anymore fwiw)), but libp2p the protocol is IMO completely fine.

If we're going the direction of getting rid of rust-libp2p, another possibility is to use the libp2p implementation of smoldot (which I wrote from scratch). It would require a ton of internal Substrate refactoring, but so would libtorrent.

Of course I'm not completely 100% happy about letting Substrate use the same code as smoldot, because one of the main objectives of smoldot is to be a _different_ implementation than Substrate. If Substrate and smoldot share too much code, it goes against this objective.

**Comment by altonen** [comment]
@burdges Yes good point

I think the problem is that to even get a hacky version of (3) working would require cooperation from two different teams and in the best case where it actually improves things enough to start actually considering it, the implementation has to be thrown away and implemented properly if it was actually a hackish implementation. I think the networking team can provide an API for libtorrent some time in the near(ish) future. After that it's up to the people working availability distribution/recovery to try out the protocol and see how it performs.

@tomaka For the past three months I've been working on a libp2p replacement and it's at a point where I can release the first real version of it. The API is very different from rust-libp2p and as such would be relatively easy to get it working in Substrate since the APIs of the library were inspired by the long-term refactoring goals we have for `sc-network` and any upper-layer protocol using `sc-network`. We still want to have libp2p as well so the actual networking stack of Substrate will be offered in a manner similar to parity-db/rocksdb. This means there will be a small set of interfaces the network backend must implement and if smoldot's network stack can implement these interfaces too, it can be ported to Substrate as well.

**Comment by zdave-parity** [comment]
FWIW I fixed the bitswap implementation on a branch here: https://github.com/zdave-parity/substrate/tree/polkadot-bulletin-chain

I rewrote it as a pretty simple custom `NetworkBehaviour`. Re the DoS concerns it's not clear to me exactly what those are, but the implementation at least has a reasonable bound on the memory consumed per connection so should not be abusable that way.

**Comment by bkontur** [comment]
@dmitry-markin Can we close this or anything needs to be done? afaik, we should support Bitswap now, right?

**Comment by dmitry-markin** [comment]
We can close this:
1. `--ipfs-server` is working now with litep2p network backend
2. I don't think it makes sense to implement Bitswap using Notifications in libp2p network backend. We can manage substreams manually, and it should be more straightforward than (ab)using notifications. The tracking issue for this is: https://github.com/paritytech/polkadot-sdk/issues/9837 (low priority).
